src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -626,15 +626,16 @@ void checkMemberStateOnConstructorExit(MethodSymbol constructor, Symbol member,
                 {
                     return;
                 }
-                if (!fieldType.NullableAnnotation.IsNotAnnotated())
+                var annotations = symbol.GetFlowAnalysisAnnotations();
+                if ((annotations & (FlowAnalysisAnnotations.AllowNull)) != 0)
```

Comment:
Is there any particular reason for introducing inconsistency between `[AllowNull]` and `[DisallowNull]` there?

I.e. `[AllowNull] string field;` mean you can left the field uninitialized
Yet `[DisallowNull] string? field;` doesn't mean you have to initialize the field

I can see that you enforced this behavior in `public void DisallowNullField_Constructor`'s `C3.F`, but I can't see why it should behave this way.

Wouldn't C# benefit from consistent compiler behavior there?

I understand that `[NotNull/DisallowNull]` and `[MaybeNull/AllowNull]` attributes will likely be used together. However in some situation a person might omit the outward annotation attributes.

Consider a scenario where a type is only created and initialized to be serialized and used in another process. In this case you might not notice that it might be useful / logical to apply `[MaybeNull]` or `[NotNull] attributes especially if you aren't very familiar with `System.Diagnostic.CodeAnalysis` annotations.
You might create a type, write initialization for it, notice warnings for some optional fields, annotate them with `[AllowNull]`, see that it works, annotate required fields with `[DisallowNull]` to make this initialization safe to refactor later and finish there.
In this case you'll apply `[AllowNull]`, see that it works and you'll only notice that `[DisallowNull]` doesn't have a similar effect only when you end up with an unexpectedly uninitialized field some years later.

Reply:
This is a reasonable point. This is pre-existing behavior, which IIRC dotnet/runtime has depended on at least in the past, where they would put AllowNull on fields and expect to have the initialization warning suppressed.

However the only suspect usage of the attribute I see now in dotnet/runtime is [here](https://github.com/dotnet/runtime/blob/31062e4ec6b248b709095e13947400b49d507e53/src/libraries/System.Private.Xml/src/System/Xml/Xsl/Xslt/CompilerScopeManager.cs#L37).

I think it's worth examining this behavior to decide whether it's still something that people need, and whether we could adjust it as a "bug-fix" level change. But I won't make the change in this PR.

Reply:
The behavior I had in mind here is enforcing initialization of fields marked with `[DisallowNull]`, not relaxing the rules around `[AllowNull]` members. This way the dotnet/runtime example you gave won't be broken and people won't have to think whether it's `[DisallowNull]` or `[NotNull]` to enforce initialization since both will work. Isn't it the reason why both `[AllowNull]` and `[MaybeNull]` were both supported originally?

Reply:
Thanks, I think I understand the suggestion fully now. However, my view is that the warnings should just work by requiring the constructor to make it so the flow state of fields is compatible with the "output state" of the field. I think if we discover that we *can't* close the hole of "AllowNull removes initialization warnings", I'm open to the concept of "DisallowNull *introduces* initialization warnings", but my first preference is definitely to make things consistent by removing the "AllowNull" hole.

Reply:
Also, because we have a specific code path here to check "are we skipping the diagnostic because of presence of AllowNull?" it should be pretty easy, if we decide it's necessary, to introduce a special diagnostic in this scenario to use in warning wave 6.

Reply:
> The behavior I had in mind here is enforcing initialization of fields marked with `[DisallowNull]`

The problem there is that can break cases where initialization cannot be done through the constructor, which is the only place that the compiler checks for it.

Reply:
@RikkiGibson FYI I've created an issue to track this since I want to get a notification when it's fixed and whether you'll change `[AllowNull]` to match `[DisallowNull]` or vice versa https://github.com/dotnet/roslyn/issues/50085